src/faiss.cc

```diff
@@ -5,80 +5,77 @@
 #include <faiss/IndexFlat.h>
 #include <faiss/index_io.h>
 #include <faiss/impl/FaissException.h>
+#include <faiss/impl/io.h>
+#include <faiss/index_factory.h>
+#include <faiss/MetricType.h>
 #include <faiss/impl/IDSelector.h>
 
 using namespace Napi;
 using idx_t = faiss::idx_t;
 
-class IndexFlatL2 : public Napi::ObjectWrap<IndexFlatL2>
+template <class T, typename Y>
+class IndexBase : public Napi::ObjectWrap<T>
 {
 public:
-  IndexFlatL2(const Napi::CallbackInfo &info) : Napi::ObjectWrap<IndexFlatL2>(info)
+  IndexBase(const Napi::CallbackInfo &info) : Napi::ObjectWrap<T>(info)
   {
     Napi::Env env = info.Env();
-    if (info[0].IsExternal())
+
+    if (!info.IsConstructCall())
```

Comment:
greatly simplified the constructors responsibilities and removed the need to pass around external data. This moves the responsibility of each feature to different functions.